### ISSUE TITLE
eth_estimateGas returns error in calls where it should succeed

### ISSUE THREAD

**Issue by almindor** [ethereum/go-ethereum#15671]
eth_estimateGas returns the new `Error: gas required exceeds allowance or always failing transaction` on calls that should succeed.

#### System information

Geth version: `Geth Version: 1.7.3-stable`
OS & Version: `Linux Ubuntu-1604-xenial-64-minimal 4.10.0-35-generic #39~16.04.1-Ubuntu SMP Wed Sep 13 09:02:42 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux`

#### Expected behaviour
web3.eth.estimateGas returns a proper gas value for a call to an ERC20 contract's Transfer method with 0 for value as well as 1 for value.

#### Actual behaviour
Works with 0 for value but returns `Error: gas required exceeds allowance or always failing transaction` if the value is set to 1.

#### Steps to reproduce the behaviour

NOTE: the following contract is [EOS token](https://etherscan.io/address/0x86fa049857e0209aa7d9e616f7eb3b3b78ecfdb0#code)
geth attach to a fully synchronized live node on mainnet and run:
```
var sender = '0xaa08abf10edbafd274668eed547f88b786a03410';
var contractAddress = '0x86fa049857e0209aa7d9e616f7eb3b3b78ecfdb0';
var receiver = '0x692a70d2e424a56d2c6c27aa97d1a86395877b3a';

var ABI = [{"constant":true,"inputs":[],"name":"name","outputs":[{"name":"","type":"bytes32"}],"payable":false,"type":"function"},{"constant":false,"inputs":[],"name":"stop","outputs":[],"payable":false,"type":"function"},{"constant":false,"inputs":[{"name":"guy","type":"address"},{"name":"wad","type":"uint256"}],"name":"approve","outputs":[{"name":"","type":"bool"}],"payable":false,"type":"function"},{"constant":false,"inputs":[{"name":"owner_","type":"address"}],"name":"setOwner","outputs":[],"payable":false,"type":"function"},{"constant":true,"inputs":[],"name":"totalSupply","outputs":[{"name":"","type":"uint256"}],"payable":false,"type":"function"},{"constant":false,"inputs":[{"name":"src","type":"address"},{"name":"dst","type":"address"},{"name":"wad","type":"uint256"}],"name":"transferFrom","outputs":[{"name":"","type":"bool"}],"payable":false,"type":"function"},{"constant":true,"inputs":[],"name":"decimals","outputs":[{"name":"","type":"uint256"}],"payable":false,"type":"function"},{"constant":false,"inputs":[{"name":"dst","type":"address"},{"name":"wad","type":"uint128"}],"name":"push","outputs":[{"name":"","type":"bool"}],"payable":false,"type":"function"},{"constant":false,"inputs":[{"name":"name_","type":"bytes32"}],"name":"setName","outputs":[],"payable":false,"type":"function"},{"constant":false,"inputs":[{"name":"wad","type":"uint128"}],"name":"mint","outputs":[],"payable":false,"type":"function"},{"constant":true,"inputs":[{"name":"src","type":"address"}],"name":"balanceOf","outputs":[{"name":"","type":"uint256"}],"payable":false,"type":"function"},{"constant":true,"inputs":[],"name":"stopped","outputs":[{"name":"","type":"bool"}],"payable":false,"type":"function"},{"constant":false,"inputs":[{"name":"authority_","type":"address"}],"name":"setAuthority","outputs":[],"payable":false,"type":"function"},{"constant":false,"inputs":[{"name":"src","type":"address"},{"name":"wad","type":"uint128"}],"name":"pull","outputs":[{"name":"","type":"bool"}],"payable":false,"type":"function"},{"constant":true,"inputs":[],"name":"owner","outputs":[{"name":"","type":"address"}],"payable":false,"type":"function"},{"constant":false,"inputs":[{"name":"wad","type":"uint128"}],"name":"burn","outputs":[],"payable":false,"type":"function"},{"constant":true,"inputs":[],"name":"symbol","outputs":[{"name":"","type":"bytes32"}],"payable":false,"type":"function"},{"constant":false,"inputs":[{"name":"dst","type":"address"},{"name":"wad","type":"uint256"}],"name":"transfer","outputs":[{"name":"","type":"bool"}],"payable":false,"type":"function"},{"constant":false,"inputs":[],"name":"start","outputs":[],"payable":false,"type":"function"},{"constant":true,"inputs":[],"name":"authority","outputs":[{"name":"","type":"address"}],"payable":false,"type":"function"},{"constant":true,"inputs":[{"name":"src","type":"address"},{"name":"guy","type":"address"}],"name":"allowance","outputs":[{"name":"","type":"uint256"}],"payable":false,"type":"function"},{"inputs":[{"name":"symbol_","type":"bytes32"}],"payable":false,"type":"constructor"},{"anonymous":true,"inputs":[{"indexed":true,"name":"sig","type":"bytes4"},{"indexed":true,"name":"guy","type":"address"},{"indexed":true,"name":"foo","type":"bytes32"},{"indexed":true,"name":"bar","type":"bytes32"},{"indexed":false,"name":"wad","type":"uint256"},{"indexed":false,"name":"fax","type":"bytes"}],"name":"LogNote","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"name":"authority","type":"address"}],"name":"LogSetAuthority","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"name":"owner","type":"address"}],"name":"LogSetOwner","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"name":"from","type":"address"},{"indexed":true,"name":"to","type":"address"},{"indexed":false,"name":"value","type":"uint256"}],"name":"Transfer","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"name":"owner","type":"address"},{"indexed":true,"name":"spender","type":"address"},{"indexed":false,"name":"value","type":"uint256"}],"name":"Approval","type":"event"}];

var contract = web3.eth.contract(ABI).at(contractAddress);

var data0 = contract.transfer.getData(receiver, 0);
var data1 = contract.transfer.getData(receiver, 1);

web3.eth.estimateGas({data: data0, from: sender, to: contractAddress});
web3.eth.estimateGas({data: data1, from: sender, to: contractAddress});
```

Result is
```
> web3.eth.estimateGas({data: data0, from: sender, to: contractAddress});
40395
> web3.eth.estimateGas({data: data1, from: sender, to: contractAddress});
Error: gas required exceeds allowance or always failing transaction
```
which does not make sense? Why would changing just the transfer value to 1 cause this error?

**Comment by karalabe** [comment]
Are you sure the CALL isn't throwing in the EVM due to perhaps the sender account not having any balance available to send? Or some similar issue?

**Comment by almindor** [comment]
You are correct, it seems some ERC20 tokens throw while others just re-value to 0.

I still think this change is a bug. Aside from not being versioned as a breaking change there are these possibilities when the error is reported:

1. The function actually ran out of gas (also low current block gaslimit might be mixed in here too)
2. The function explicitly threw
3. The data to the call is malformed

Is there any way these three could be reported as different errors/codes? It is otherwise quite hard to handle on client side.

**Comment by veox** [comment]
> Is there any way these three could be reported as different errors/codes?

The linked contract has been compiled with `solc v0.4.11+commit.68ef5810`. A cursory glance at the sources seems to indicate that it's not using `revert` anywhere (although it should have technically been [available since v0.4.10](https://github.com/ethereum/solidity/releases/tag/v0.4.10)), only `assert` and `throw`.

So, even if `geth` was capable of discerning between the three cases you outline (I don't know if it already does), there's no way to do so for the particular linked contract.
